server/implementation/src/main/java/io/smallrye/graphql/execution/datafetcher/decorator/ValidationDecorator.java

```diff
@@ -0,0 +1,155 @@
+package io.smallrye.graphql.execution.datafetcher.decorator;
+
+import static graphql.ErrorType.ValidationError;
+import static java.util.Collections.singletonList;
+import static java.util.stream.Collectors.toList;
+
+import java.lang.annotation.Annotation;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import javax.validation.ConstraintViolation;
+import javax.validation.Path.Node;
+import javax.validation.Valid;
+import javax.validation.Validation;
+import javax.validation.ValidatorFactory;
+import javax.validation.groups.ConvertGroup;
+import javax.validation.groups.Default;
+
+import graphql.ErrorClassification;
+import graphql.GraphQLError;
+import graphql.execution.DataFetcherResult.Builder;
+import graphql.language.Argument;
+import graphql.language.SourceLocation;
+import graphql.schema.DataFetchingEnvironment;
+import graphql.schema.GraphQLArgument;
+import io.smallrye.graphql.execution.datafetcher.ExecutionContext;
+import io.smallrye.graphql.transformation.TransformException;
+
+public class ValidationDecorator extends AbstractDataFetcherDecorator {
+    private static final ValidatorFactory VALIDATOR_FACTORY = Validation.buildDefaultValidatorFactory();
+
+    @Override
+    protected void before(ExecutionContext executionContext) throws TransformException {
+        Annotation[][] parametersAnnotations = executionContext.method().getParameterAnnotations();
+        for (int i = 0; i < parametersAnnotations.length; i++) {
+            Annotation[] annotations = parametersAnnotations[i];
+            for (Annotation annotation : annotations) {
+                if (Valid.class.equals(annotation.annotationType())) {
+                    Object argument = executionContext.arguments()[i];
+
+                    // TODO validate annotations on the parameter itself
+                    Set<ConstraintViolation<Object>> violations = VALIDATOR_FACTORY.getValidator().validate(argument,
+                            groups(annotations));
+
+                    if (!violations.isEmpty()) {
+                        int parameterIndex = i;
+                        throw new TransformException(null, null, null) {
+                            @Override
+                            public Builder<Object> appendDataFetcherResult(Builder<Object> builder,
+                                    DataFetchingEnvironment dfe) {
+                                violations.forEach(violation -> builder
+                                        .error(new ValidationFailedGraphQLError(dfe, violation, parameterIndex)));
+                                return builder;
+                            }
```

Comment:
The BV exceptions don't transport enough details for the `ReflectionDataFetcher` to merge into the error result. And it does that already with `TransformationException`s, so I thought it would be leaner to hook in that mechanism.

Reply:
Yea, but it's not a transform exception ? Should we not then create a new Exception (similar to TransformException) like ValidationException ?

Reply:
Or maybe a common super class or interface for `TransformException` and `ValidationException`? Something like `DataFetchingException`?

Reply:
Yea that should work

Reply:
Like so? (see my latest commit)

Reply:
I like it, one last ask.
1) The package seems wrong. maybe it should be in `io.smallrye.graphql.execution`.
2) There is already a `io.smallrye.graphql.execution.datafetcher.DataFetcherException`, so maybe it should be named something else ? I am just not sure what though ...

Reply:
I suggest moving it to `io.smallrye.graphql.execution` and maybe rename to `AbstractFetcherException' or `AbstractDataFetcherException'

Reply:
I'm not deep enough into that code to understand what the semantics of the `datafetcher` and `decorator` packages and their classes. So I wonder if it's consistent to have a dependency from `io.smallrye.graphql.execution.validation` or so down into `io.smallrye.graphql.execution.datafetcher.decorator`.

Reply:
I think that is fine.

Reply:
> I think that is fine.

I'm not sure what exactly you mean :-)

Reply:
I think that is fine .... to have a dependency from io.smallrye.graphql.execution.validation or so down into io.smallrye.graphql.execution.datafetcher.decorator.

Reply:
Like so?